sdk/core/System.ClientModel/src/Internal/SSE/AsyncServerSentEventEnumerable.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Internal;
+
+/// <summary>
+/// Represents a collection of SSE events that can be enumerated as a C# async stream.
+/// </summary>
+internal class AsyncServerSentEventEnumerable : IAsyncEnumerable<ServerSentEvent>
+{
+    private readonly Stream _contentStream;
+
+    public AsyncServerSentEventEnumerable(Stream contentStream)
+    {
+        _contentStream = contentStream;
+    }
+
+    public IAsyncEnumerator<ServerSentEvent> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+    {
+        return new AsyncServerSentEventEnumerator(_contentStream, cancellationToken);
```

Comment:
I know that we generally don't make model types thread-safe, but it is worth asking if it's a concern in this case.

Reply:
My sense is that it's reasonable to let these follow the [IAsyncEnumerable pattern for thread-safety](https://learn.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8#much-ado-about-threading), but let me know if you feel differently.

Reply:
I think that's reasonable, though it talks about having multiple threads call `MoveNext`.   What I'm thinking of is the scenario where two different threads call `GetAsyncEnumerator`.  They would get independent enumerator instances to work with, avoiding the "don't call MoveNext from two threads" guidance, but since both enumerators are backed by the same `_contentStream`, they would still not be safe to use concurrently.

To be clear:  I'm not advocating for thread safety, just making sure that we're good with the trade off and have considered whether or not to add some kind of guard/throw when there's already one active enumerator using the stream.

Reply:
GetAsyncEnumerator should only be called from a client-implemented IAsyncEnumerable type or by code written by the C# compiler to implement the type backing the `await foreach`.  If your specific concern is these scenarios, it feels reasonable to say that a client-author implementing an SSE enumerator should know it's not safe to read concurrently from the SSE stream and the C# compiler should do the right thing in all cases.